### PR TITLE
BUG: Turn off periodicity for particle datasets when a bounding box is explicitly set

### PR DESCRIPTION
**Pull request by jzuhone**
## PR Summary

In Issue #2639 it was reported that yt selection routines return too may particles when a bounding box is explicitly applied to a particle dataset with periodicity. 

In the comment thread, it was noted that at least for now it makes sense to turn off periodicity when a bounding box is applied, which this PR does. 

Closes (in a sense) Issue #2639.

## PR Checklist

- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

File: yt/data_objects/static_output.py
Comment (neutrinoceros) [inline yt/data_objects/static_output.py:1946]: I believe this should solve the failures
```suggestion
        self._periodicity = (False,) * 3
```

**Review by neutrinoceros — COMMENTED** [review]
Thank you job for patching this.

File: yt/data_objects/static_output.py
Comment (jzuhone) [reply yt/data_objects/static_output.py]: Whoops, thanks

**Review by matthewturk — APPROVED** [review]
I think this is correct.  I see no reason not to merge immediately.

**Comment by jzuhone** [comment]
@neutrinoceros Wait on merging until I check something real quick

**Review by AshKelly — APPROVED** [review]

**Comment by jzuhone** [comment]
Ok, I did this in a hurry last night and I wasn't careful enough. I think that in some cases this new function gets called _before_  `_parse_parameter_file` sets the periodicity itself, so I need to fix that before this can go in.

Sorry about this.

**Comment by neutrinoceros** [comment]
> Ok, I did this in a hurry last night and I wasn't careful enough. I think that in some cases this new function gets called before _parse_parameter_file sets the periodicity itself, so I need to fix that before this can go in.
Sorry about this.

No problem, we share a responsibility in reviewing this in a hurry too. I think we're all a bit shaky with the perspective of the upcoming release. Could you switch to draft mode and back to give a clear signal when this is ready for review again ?

**Comment by jzuhone** [comment]
If the tests pass it's good

File: yt/frontends/sdf/data_structures.py
Comment (neutrinoceros) [inline yt/frontends/sdf/data_structures.py:122]: I think that this will wrongfully emit a warning even in case the `else` right before this is hit. IMO this conditional should be included in the previous one.

**Review by neutrinoceros — CHANGES_REQUESTED** [review]
thanks for being more cautious ! I think there's one more tiny thing to improve here but otherwise this looks good.

File: yt/frontends/sdf/data_structures.py
Comment (jzuhone) [reply yt/frontends/sdf/data_structures.py]: This is fixed in the latest commit--thanks for pointing it out

File: yt/frontends/sdf/data_structures.py
Comment (neutrinoceros) [inline yt/frontends/sdf/data_structures.py:117]: Could we make idioms consistent here while we're at it ?
```suggestion
        if self.parameters.get("do_periodic", False):
```

**Review by neutrinoceros — DISMISSED** [review]

**Comment by jzuhone** [comment]
So it looks like we'd have to fix this by modifying `sph_answer`--the Arepo test fails because one of the things it checks involves getting a region which goes outside of the bounding box, and of course our typical solution to that is to _enable_ periodicity. Not sure what to do about this now. 

Don't let this hold up 4.0.

**Comment by neutrinoceros** [comment]
ok, just going to use the "do not merge" label right now to prevent accidents since we have THREE approvals (mine included  🙈) on a PR that's not ready yet.
Thanks John for your patience, I hope we can solve this soon :)

**Comment by jzuhone** [comment]
I'm resurrecting this PR and I think I'll need to fix the Arepo answer tests.

**Comment by neutrinoceros** [comment]
requesting a review from myself, I'll look over this after CI completes !

File: yt/data_objects/static_output.py
Comment (neutrinoceros) [inline yt/data_objects/static_output.py:1976]:
```suggestion
        # user--previously in these circumstances periodicity behaved
```

File: yt/data_objects/static_output.py
Comment (neutrinoceros) [inline yt/data_objects/static_output.py:1983]: I guess the warning should only be emitted if the method has an effect (i.e. if periodicity isn't already False)

File: yt/data_objects/static_output.py
Comment (jzuhone) [reply yt/data_objects/static_output.py:1976]: It's still true that periodicity behaves oddly if we apply a bounding box--I don't think the change in tense is what we want here.

Comment (neutrinoceros) [reply yt/data_objects/static_output.py:1976]: ah I see, nevermind

**Comment by neutrinoceros** [comment]
note: the docs build failed with 
```
ERROR: Could not build wheels for yt, which is required to install pyproject.toml-based projects
```
this is unrelated to the PR and caused by newest versions of setuptools, see https://github.com/pypa/setuptools/issues/3322

**Comment by jzuhone** [comment]
@Xarthisius I think the Jenkins test are failing because of Issue #3327

**Comment by Xarthisius** [comment]
> @Xarthisius I think the Jenkins test are failing because of Issue #3327

I'm trying to come up with something that wouldn't be a fully-fledged solution like redesigning ewah storage format, and on the other hand, wouldn't rely on removing all ewah files for each test run. How about allowing to ignore existing ewah as a kwarg to load?

**Comment by jzuhone** [comment]
> I'm trying to come up with something that wouldn't be a fully-fledged solution like redesigning ewah storage format, and on the other hand, wouldn't rely on removing all ewah files for each test run. 

Do you need to remove all of the files, or just the ones being used in the particular PR? 

> How about allowing to ignore existing ewah as a kwarg to load? Another option: appending unique_id to ewah filename?

I think either of these solutions is acceptable.

**Comment by Xarthisius** [comment]
> Do you need to remove all of the files, or just the ones being used in the particular PR?

From the tests' perspective I don't know which files are affected without doing some advanced parsing of diffs or something similar. Also I'd have to do it always, not once. If I removed ewah e.g. for your PR only, new one would be generated and it would break all the other test runs until your PR is merged.

> > How about allowing to ignore existing ewah as a kwarg to load? Another option: appending unique_id to ewah filename?
> 
> I think either of these solutions is acceptable.

I'll try to issue a PR in a (hopefully) near future.

**Comment by neutrinoceros** [comment]
> How about allowing to ignore existing ewah as a kwarg to load? Another option: appending unique_id to ewah filename?

For the record, I have a preference for the second solution. Adding kwargs to `load` is a last resort IMO.

**Comment by neutrinoceros** [comment]
In #3930 I'm trying to overhaul `unique_identifier`'s implementation (reducing duplication, mostly). Doing so I see that most frontends implement a reproducible id (only depends on the input filename), but some don't, and default to using `time.time()`. Is this an obstacle to your second solution @Xarthisius ? Are there downsides to changing/simplifying the implementations, like additional maintenance ?

**Comment by Xarthisius** [comment]
> In #3930 I'm trying to overhaul `unique_identifier`'s implementation (reducing duplication, mostly). Doing so I see that most frontends implement a reproducible id (only depends on the input filename), but some don't, and default to using `time.time()`. Is this an obstacle to your second solution @Xarthisius ? Are there downsides to changing/simplifying the implementations, like additional maintenance ?

I wasn't planning on using Datasets' `unique_identifier`. As you noted, it's mostly based on the original filename and it doesn't capture the info we need to store in ewah to avoid collisions (like bounding box).

**Comment by neutrinoceros** [comment]
Oh I see, thanks for the clarification !

**Comment by neutrinoceros** [comment]
With #3916 in mind, I wonder if the present patch is acceptable. Doesn't it make it completely impossible to work with a bounding box spanning over multiple corners of a periodic domain ?

**Comment by matthewturk** [comment]
@neutrinoceros I think in those particular cases we should probably allow for a translation argument.  But is it the case that a bounding box that spans multiple corners will already run up against issues?  I think it may very well be.

**Comment by neutrinoceros** [comment]
> But is it the case that a bounding box that spans multiple corners will already run up against issues?

Most likely yeah, since there are already problems without a bounding box, as we can see in https://github.com/yt-project/yt/issues/3916

I'm worried about trying to solve _that_ issue only to get locked out of solution because of the present PR.

**Comment by jzuhone** [comment]
@neutrinoceros Once you've specified a bounding box that is potentially different from the box that the set of particles originally came from, the whole notion of periodicity goes out the window. Especially if you cross only one axis. This solution is needed currently to make datasets like this behave even somewhat normally under these circumstances until we come up with a better strategy.

File: yt/data_objects/static_output.py
Comment (neutrinoceros) [inline yt/data_objects/static_output.py:1992]:
```suggestion
        if any(self.periodicity):
```

File: yt/data_objects/static_output.py
Comment (neutrinoceros) [inline yt/data_objects/static_output.py:1997]: using the encapsulated form will ease refactoring. It also sets the `self._force_periodicity` flag. 
```suggestion
            self.force_periodicity(False)
```

**Review by neutrinoceros — COMMENTED** [review]
Ok thanks for clarifying. I think the better (long term) solution would be to have Dataset objects store boundary conditions for each direction/side, rather than a single boolean per direction for periodicity.
In that framework, it should be possible to set special boundaries for when a bounding box is set.

One last question: how do you know you've updated all frontends that needed this change ?

File: yt/data_objects/static_output.py
Comment (neutrinoceros) [inline yt/data_objects/static_output.py:1997]: actually, this logic could just be encapsulated inside the `periodicity` property, couldn't it ?
Then you wouldn't need to update anything else, as the change would already affect all frontends !

Comment (jzuhone) [reply yt/data_objects/static_output.py:1997]: You don't want to change all frontends--only the ones that accept a `bounding_box` argument

Comment (neutrinoceros) [reply yt/data_objects/static_output.py:1997]: Aren't those the same that use the optional attribute `_domain_override` ?

Comment (jzuhone) [reply yt/data_objects/static_output.py:1997]: Yes, but I'm not sure what you are getting at here.

Comment (neutrinoceros) [reply yt/data_objects/static_output.py:1997]: What I mean is that it seems to me that all of this could be accomplished with the `Dataset.periodicity` property definition using a `if getattr(self, "_domain_override", False):` clause. Is this incorrect ?

Comment (jzuhone) [reply yt/data_objects/static_output.py:1997]: Ah, ok--yes, that's correct.

Comment (neutrinoceros) [reply yt/data_objects/static_output.py:1997]: Here's exactly what I have in mind
```patch
diff --git a/yt/data_objects/static_output.py b/yt/data_objects/static_output.py
index 96d013992..328f66262 100644
--- a/yt/data_objects/static_output.py
+++ b/yt/data_objects/static_output.py
@@ -322,6 +322,14 @@ class Dataset(abc.ABC):
     def periodicity(self):
         if self._force_periodicity:
             return (True, True, True)
+        elif hasattr(self, "_domain_override"):
+            # dataset loaded with a bounding box
+            if any(self._periodicity):
+                mylog.warning(
+                    "A bounding box was explicitly specified, so we "
+                    "are disabling periodicity."
+                )
+            return (False, False, False)
         return self._periodicity

     def force_periodicity(self, val=True):
```

Comment (jzuhone) [reply yt/data_objects/static_output.py:2055]: Ok I've added it

**Comment by neutrinoceros** [comment]
@jzuhone is this ready ? I'll take one last look again tomorrow but if you're done with it I might just merge it then

**Comment by jzuhone** [comment]
I recognize now that the `_disable_periodicity_for_bbox` method is unnecessary, but I will wait to remove it until after I see the tests pass--a bit worried about the potential answer test failure due to the wrong EWAH file. Will know in a few minutes...

**Comment by jzuhone** [comment]
@Xarthisius so this PR will keep failing the `halo_59` answer tests because of https://github.com/yt-project/yt/issues/3327. Any idea what we can do about it?

**Comment by Xarthisius** [comment]
> @Xarthisius so this PR will keep failing the `halo_59` answer tests because of #3327. Any idea what we can do about it?

Sans fixing the issue you've mentioned, we could remove ewah file every time the test is run.

**Comment by jzuhone** [comment]
So it turns out that this is fixed by simply centering the bounding box of the AREPO dataset on the halo center. The EWAH file is a non-issue here.

**Comment by neutrinoceros** [comment]
@jzuhone how can we check that the test isn't flaky ?

**Comment by jzuhone** [comment]
@neutrinoceros The reason the test was having so much trouble before is that I had originally changed the bounding box in such a way that one of the sphere tests was landing in a part of the domain where there was no data (I did not realize this originally). Now I have placed the halo in the center of the domain and the immediate result was that the test failed because the answers were not identical, which is precisely what we should we should expect if we change the bounding box. I looked at the projection difference images and it showed that the object merely shifted, which is also what I expected. So I am pretty confident in the test now.

**Comment by neutrinoceros** [comment]
Alright that make sense, thank you for this summary ! I'll run the test a couple times more just in case
@yt-fido tes this please

**Comment by jzuhone** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by jzuhone** [comment]
@yt-fido test this please

**Comment by jzuhone** [comment]
Ok, we ran the tests twice more--this is good to go.

**Review by neutrinoceros — COMMENTED** [review]
I think we should add a couple lines to test that `ds.periodicity`  is correctly outputted with and without the bounding box, to avoid future regressions. Other than that LGTM

File: yt/data_objects/static_output.py
Comment (neutrinoceros) [inline yt/data_objects/static_output.py:329]: This warning might trigger at surprising times because the periodicity property is often accessed internally. It would probably be preferable to issue the warning when the bounding box is created instead.

**Review by neutrinoceros — APPROVED** [review]

**Pull request merged**
